File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] ans = new int[nums.length];

        Stack<Integer>st = new Stack<>();
        for(int i=nums.length-1;i>=0;i--){
            st.add(nums[i]);
        }

        for(int i=nums.length-1;i>=0;i--){
           while(st.size() > 0 && st.peek() <= nums[i])st.pop();
           if(st.size() == 0)ans[i] = -1;
           else ans[i] = st.peek();
           st.add(nums[i]);
        }
        return ans;
    }
}

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
         stack<int>st;
         vector<int>ans(nums.size(),-1);
         int n = nums.size();
        for(int i =(2*nums.size())-1; i>=0;i--){
            
            while(!st.empty() && st.top()<=nums[i%n]){
             st.pop();
            }
            if(!st.empty()){
                ans[i%n] = st.top();
            }else{
                ans[i%n] = -1;
            }
            st.push(nums[i%n]);
        }

        return ans;
    }
};